### DISCUSSION TITLE
KVS - Limitations and behaviors of Keys

### DISCUSSION THREAD

**Discussion by keithpine** [Seagate/cortx-motr#1234] [Q&A] [answered]
I have some basic questions regarding the Motr Key-Value Store, particularly in regards to the keys:

1. Is there a maximum length imposed on either the key or value?
2. Does the key length affect the performance of KVS operations (lookup/insert/delete) at all?
3. Is there a recommended key size (e.g. <= 64 bytes), to optimize performance and storage?
4. What is the sorting/comparison order of the keys? My understanding is that, for something like a string key, it's a lexicographical order. Is that just done with unsigned byte comparisons (e.g. memcmp)?
5. Is it possible for a client to define a custom key comparison? For example, if I was using strings as keys, could I define a sorting function to store the keys in reverse order, so that NEXT key operations would return the keys in that reverse order? I see some ability to define this in Motr (`m0_be_btree_kv_ops.ko_compare`), but I assume that's internal and not available at the application layer.

**Comment by andriytk** [comment, marked as answer]
Hi @keithpine, let me try to answer. cc @nikitadanilov, @madhavemuri, @just-now, @ivan-alekhin, @huanghua78.

1. I think, the limit is imposed by the BE (back-end) maximum transaction size for key+value at the server side, which is configured with these parameters: https://github.com/Seagate/cortx-motr/blob/47e7b4aa1d87588cde7aed517e35dc0eb2e2c870/scripts/install/etc/sysconfig/motr#L87-L88
2. Yes, limited by memcmp(k1, k2, len) performance on log(n) records in the index.
3. Not that I'm aware of.
4. See here: https://github.com/Seagate/cortx-motr/blob/47e7b4aa1d87588cde7aed517e35dc0eb2e2c870/motr/client.h#L369-L372
5. No, I'm afraid such capability is not exposed on the Motr API level.
